pkg/sgxcontroller/sgxcontroller.go

```diff
@@ -0,0 +1,101 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sgxcontroller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/internal/containers"
+)
+
+// +kubebuilder:webhook:path=/pods-sgx,mutating=true,failurePolicy=fail,groups="",resources=pods,verbs=create;update,versions=v1,name=sgx.mutator.webhooks.intel.com,sideEffects=NoneOnDryRun
+
+// SgxMutator annotates Pods.
+type SgxMutator struct {
+	Client  client.Client
+	decoder *admission.Decoder
+}
+
+const (
+	namespace           = "sgx.intel.com"
+	encl                = namespace + "/enclave"
+	epc                 = namespace + "/epc"
+	provision           = namespace + "/provision"
+	provisionAnnotation = namespace + "/needs-provision"
+)
+
+func (s *SgxMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+
+	err := s.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	totalEpc := int64(0)
+
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+	for idx, container := range pod.Spec.Containers {
+		requestedResources, err2 := containers.GetRequestedResources(container, namespace)
+		if err2 != nil {
+			return admission.Errored(http.StatusInternalServerError, err2)
+		}
+		if _, ok := requestedResources[encl]; ok {
+			return admission.Denied(fmt.Sprintf("%s resource requests not allowed", encl))
+		}
+		if _, ok := requestedResources[provision]; ok {
+			return admission.Denied(fmt.Sprintf("%s resource requests not allowed", provision))
+		}
+		if epcsize, ok := requestedResources[epc]; ok {
+			totalEpc += epcsize
+
+			attestation, found := pod.Annotations[provisionAnnotation]
+			if found && attestation == "yes" {
+				pod.Spec.Containers[idx].Resources.Limits[corev1.ResourceName(provision)] = resource.MustParse("1")
+				pod.Spec.Containers[idx].Resources.Requests[corev1.ResourceName(provision)] = resource.MustParse("1")
+			}
+			pod.Spec.Containers[idx].Resources.Limits[corev1.ResourceName(encl)] = resource.MustParse("1")
+			pod.Spec.Containers[idx].Resources.Requests[corev1.ResourceName(encl)] = resource.MustParse("1")
+		}
+	}
+
+	if totalEpc != 0 {
+		quantity := resource.NewQuantity(totalEpc, resource.BinarySI)
+		pod.Annotations["sgx.intel.com/epc"] = quantity.String()
+	}
+
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+// InjectDecoder injects the decoder.
+func (s *SgxMutator) InjectDecoder(d *admission.Decoder) error {
```

Comment:
Is it an implementation of some interface? Could you please reference it in the comment?

Reply:
fixed!